File: chorum_murohc/api/rewards.py
"""Reward-catalogue endpoints: the household's list of rewards a child may
redeem, and its management.

Five routes under `/api/v1/rewards/`, shaped exactly like
`chorum_murohc/api/chores.py`'s chore-pool routes, same-origin session
authenticated and scoped to the one household the caller is a live member of:

- `GET rewards/` lists rewards. A child sees `id`, `name` and `points` for
  the active rewards only. A parent sees the full management shape, active
  by default and both states behind `?include_inactive=true`.
- `POST rewards/` creates a reward. Parent only.
- `GET`, `PATCH` and `DELETE rewards/<pk>/` read, edit and remove one reward.
  Parent only.
- `POST rewards/<pk>/deactivate/` and `POST rewards/<pk>/reactivate/` move a
  reward between the two states. Parent only.

Authority is never taken from the client. The household comes from
`resolve_active_membership`, never from a body, query string, path or
header, and `get_queryset()` filters by it before any lookup, so a reward of
another household is an ordinary generic 404 rather than a denial that
confirms the row exists. Every mutation re-resolves the actor and re-reads
the row inside one transaction, and writes its audit event in that same
transaction.

Redeeming a reward, and a parent's fulfil or cancel of that redemption, are
not here: they read and write the ledger and live in
`chorum_murohc/api/redemptions.py`, on top of `chorum_murohc.rewards.services`.
"""

from django.db import IntegrityError, transaction
from rest_framework import serializers, status
from rest_framework.authentication import SessionAuthentication
from rest_framework.exceptions import NotFound, PermissionDenied
from rest_framework.renderers import JSONRenderer
from rest_framework.response import Response
from rest_framework.views import APIView

from chorum_murohc.api.permissions import (
    PERMISSION_DENIED_DETAIL,
    IsHouseholdChild,
    IsHouseholdParent,
)
from chorum_murohc.api.session import resolve_active_membership
from chorum_murohc.audit.models import AuditEvent
from chorum_murohc.identity.models import Membership
from chorum_murohc.rewards.models import Reward

AUDIT_TARGET_TYPE = 'reward'
AUDIT_CREATE = 'reward.create'
AUDIT_UPDATE = 'reward.update'
AUDIT_DEACTIVATE = 'reward.deactivate'
AUDIT_REACTIVATE = 'reward.reactivate'
AUDIT_DELETE = 'reward.delete'

# The largest value the `points` column can hold. Anything above it is a
# validation error rather than a database error.
POINTS_MAXIMUM = 2147483647

# The one detail returned when a name already exists in the caller's own
# household. It names no other household and echoes no stored row.
DUPLICATE_NAME_DETAIL = 'A reward with this name already exists in this household.'

# The database index behind the case-insensitive per-household name rule. Its
# violation is the same 400 as the serializer's, never a 500.
NAME_UNIQUE_CONSTRAINT = 'reward_hh_name_ci_unique'

# The exact query string that widens a parent list to both states.
INCLUDE_INACTIVE_PARAMETER = 'include_inactive'
INCLUDE_INACTIVE_VALUE = 'true'


# Listing the catalogue is the one route a child shares with a parent.
IsHouseholdParentOrChild = IsHouseholdParent | IsHouseholdChild


def duplicate_name_exists(household, name):
    """Report a reward of `household` whose name matches `name` ignoring case."""
    return Reward.objects.filter(household=household, name__iexact=name).exists()


class RewardChildSerializer(serializers.ModelSerializer):
    """What a child may see: three fields, and never a state or a timestamp."""

    class Meta:
        model = Reward
        fields = ('id', 'name', 'points')
        read_only_fields = fields


class RewardParentSerializer(serializers.ModelSerializer):
    """What a parent may see: the same three fields plus the state and times."""

    class Meta:
        model = Reward
        fields = ('id', 'name', 'points', 'is_active', 'created_at', 'updated_at')
        read_only_fields = fields


class RewardWriteSerializer(serializers.ModelSerializer):
    """The only writable body: exactly a name and a point cost.

    Everything else a caller may send is ignored, so a body carrying
    `household`, `id`, `is_active` or a timestamp cannot place a reward in
    another household, choose its identifier, or bypass the two named state
    actions. The household to check names against arrives in the context, not
    in the body.
    """

    class Meta:
        model = Reward
        fields = ('name', 'points')
        extra_kwargs = {  # noqa: RUF012
            'name': {'required': True, 'allow_blank': False},
            'points': {'required': True, 'min_value': 1, 'max_value': POINTS_MAXIMUM},
        }

    def validate_name(self, value):
        if self.instance is not None and value == self.instance.name:
            return value
        if duplicate_name_exists(self.context['household'], value):
            raise serializers.ValidationError(DUPLICATE_NAME_DETAIL)
        return value


def duplicate_name_error(error):
    """Turn the name index violation into the same 400 the serializer returns."""
    if NAME_UNIQUE_CONSTRAINT not in str(error):
        raise error
    return serializers.ValidationError({'name': [DUPLICATE_NAME_DETAIL]})


class _RewardAPIView(APIView):
    """Shared authority, scoping and audit behaviour for the reward routes."""

    renderer_classes = (JSONRenderer,)
    authentication_classes = (SessionAuthentication,)
    permission_classes = (IsHouseholdParent,)
    permission_household_attribute = 'household'

    def initial(self, request, *args, **kwargs):
        if request.method.lower() not in self.http_method_names:
            self.http_method_not_allowed(request)
        super().initial(request, *args, **kwargs)

    def get_permission_household(self, request):
        membership = resolve_active_membership(request.user)
        return None if membership is None else membership.household

    def get_queryset(self):
        household = self.get_permission_household(self.request)
        if household is None:
            return Reward.objects.none()
        return Reward.objects.filter(household=household)

    def require_parent_membership(self):
        membership = resolve_active_membership(self.request.user)
        if membership is None or membership.role != Membership.Role.PARENT:
            raise PermissionDenied(PERMISSION_DENIED_DETAIL)
        return membership

    def get_reward(self, pk, *, lock=False):
        household = self.get_permission_household(self.request)
        rewards = Reward.objects.filter(household=household)
        if lock:
            rewards = rewards.select_for_update()
        reward = rewards.filter(pk=pk).first()
        if reward is None:
            raise NotFound()
        self.check_object_permissions(self.request, reward)
        return reward

    def write_audit_event(self, membership, action, reward_id, context):
        AuditEvent.objects.create(
            household=membership.household,
            actor=membership.user,
            action=action,
            target_type=AUDIT_TARGET_TYPE,
            target_id=str(reward_id),
            context=context,
        )


class RewardListView(_RewardAPIView):
    """`GET /api/v1/rewards/` and `POST /api/v1/rewards/`."""

    http_method_names = ('get', 'head', 'post', 'options')

    def get_permissions(self):
        if self.request.method == 'POST':
            return [IsHouseholdParent()]
        return [IsHouseholdParentOrChild()]

    def get(self, request):
        membership = resolve_active_membership(request.user)
        is_parent = membership is not None and membership.role == Membership.Role.PARENT
        rewards = self.get_queryset()

        include_inactive = is_parent and (
            request.query_params.get(INCLUDE_INACTIVE_PARAMETER)
            == INCLUDE_INACTIVE_VALUE
        )
        if not include_inactive:
            rewards = rewards.filter(is_active=True)

        serializer_class = (
            RewardParentSerializer if is_parent else RewardChildSerializer
        )
        return Response(serializer_class(rewards, many=True).data)

    def post(self, request):
        try:
            with transaction.atomic():
                membership = self.require_parent_membership()
                serializer = RewardWriteSerializer(
                    data=request.data,
                    context={'household': membership.household},
                )
                serializer.is_valid(raise_exception=True)
                reward = Reward.objects.create(
                    household=membership.household,
                    name=serializer.validated_data['name'],
                    points=serializer.validated_data['points'],
                    is_active=True,
                )
                self.write_audit_event(
                    membership,
                    AUDIT_CREATE,
                    reward.pk,
                    {
                        'actor_id': membership.user_id,
                        'reward_id': reward.pk,
                        'name': reward.name,
                        'points': reward.points,
                        'is_active': reward.is_active,
                    },
                )
                body = RewardParentSerializer(reward).data
        except IntegrityError as error:
            raise duplicate_name_error(error) from error
        return Response(body, status=status.HTTP_201_CREATED)


class RewardDetailView(_RewardAPIView):
    """`GET`, `PATCH` and `DELETE /api/v1/rewards/<pk>/`, parent only."""

    http_method_names = ('get', 'head', 'patch', 'delete', 'options')

    def get(self, request, pk):
        return Response(RewardParentSerializer(self.get_reward(pk)).data)

    def patch(self, request, pk):
        try:
            with transaction.atomic():
                membership = self.require_parent_membership()
                reward = self.get_reward(pk, lock=True)
                serializer = RewardWriteSerializer(
                    reward,
                    data=request.data,
                    partial=True,
                    context={'household': membership.household},
                )
                serializer.is_valid(raise_exception=True)

                changes = {}
                for field, new_value in serializer.validated_data.items():
                    old_value = getattr(reward, field)
                    if old_value != new_value:
                        changes[field] = {'before': old_value, 'after': new_value}
                        setattr(reward, field, new_value)

                if changes:
                    reward.save(update_fields=(*changes, 'updated_at'))
                    self.write_audit_event(
                        membership,
                        AUDIT_UPDATE,
                        reward.pk,
                        {
                            'actor_id': membership.user_id,
                            'reward_id': reward.pk,
                            'changes': changes,
                        },
                    )
                body = RewardParentSerializer(reward).data
        except IntegrityError as error:
            raise duplicate_name_error(error) from error
        return Response(body)

    def delete(self, request, pk):
        with transaction.atomic():
            membership = self.require_parent_membership()
            reward = self.get_reward(pk, lock=True)
            context = {
                'actor_id': membership.user_id,
                'reward_id': reward.pk,
                'points': reward.points,
                'is_active': reward.is_active,
            }
            reward_id = reward.pk
            reward.delete()
            self.write_audit_event(membership, AUDIT_DELETE, reward_id, context)
        return Response(status=status.HTTP_204_NO_CONTENT)


class _RewardStateView(_RewardAPIView):
    """One parent-only state transition, written exactly once."""

    http_method_names = ('post', 'options')
    target_state = None
    audit_action = None

    def post(self, request, pk):
        with transaction.atomic():
            membership = self.require_parent_membership()
            reward = self.get_reward(pk, lock=True)
            if reward.is_active != self.target_state:
                previous_state = reward.is_active
                reward.is_active = self.target_state
                reward.save(update_fields=('is_active', 'updated_at'))
                self.write_audit_event(
                    membership,
                    self.audit_action,
                    reward.pk,
                    {
                        'actor_id': membership.user_id,
                        'reward_id': reward.pk,
                        'is_active': {
                            'before': previous_state,
                            'after': self.target_state,
                        },
                    },
                )
            body = RewardParentSerializer(reward).data
        return Response(body)


class RewardDeactivateView(_RewardStateView):
    """`POST /api/v1/rewards/<pk>/deactivate/`: take a reward out of use."""

    target_state = False
    audit_action = AUDIT_DEACTIVATE


class RewardReactivateView(_RewardStateView):
    """`POST /api/v1/rewards/<pk>/reactivate/`: put a reward back into use."""

    target_state = True
    audit_action = AUDIT_REACTIVATE

File: frontend/src/api/rewards.ts
/**
 * The reward-catalogue API client, for the merged `/api/v1/rewards/` routes
 * (issue #55). Shaped exactly like `api/chores.ts`, which those routes
 * mirror on the server: same-origin session authentication and CSRF only,
 * `ensureCsrfToken` and its header reused rather than duplicated, and every
 * response validated at runtime before it is trusted.
 *
 * `fetchRewards` and every write belong to the parent reward-management
 * screen, so they read and send the parent shape (id, name, points,
 * is_active, created_at, updated_at). `fetchChildRewards` is the separate,
 * narrower call the child rewards screen uses instead: it validates the
 * three-field shape the server actually sends a child, exactly as
 * `api/chores.ts` reserves its own six-field validator for the parent alone.
 */

import { CSRF_HEADER_NAME, ensureCsrfToken } from './session'

export interface Reward {
  id: number
  name: string
  points: number
  is_active: boolean
  created_at: string
  updated_at: string
}

export const REWARDS_QUERY_KEY_ROOT = 'rewards' as const

export const rewardsQueryKey = (includeInactive: boolean) =>
  [REWARDS_QUERY_KEY_ROOT, { includeInactive }] as const

/**
 * - `validation`: the request body was refused. `fieldErrors` carries the
 *   first message per field the server named.
 * - `forbidden`: the caller is no longer an active parent of this household.
 * - `notFound`: the reward does not exist, or belongs to another household,
 *   or was already removed by another session.
 * - `unavailable`: unreachable, timed out, not JSON, or an unexpected status.
 */
export type RewardFailureKind = 'validation' | 'forbidden' | 'notFound' | 'unavailable'

export class RewardRequestError extends Error {
  readonly kind: RewardFailureKind
  readonly fieldErrors: Readonly<Record<string, string>>

  constructor(kind: RewardFailureKind, fieldErrors: Record<string, string> = {}) {
    super(`Reward request failed: ${kind}`)
    this.name = 'RewardRequestError'
    this.kind = kind
    this.fieldErrors = fieldErrors
  }
}

const isRecord = (value: unknown): value is Record<string, unknown> =>
  typeof value === 'object' && value !== null && !Array.isArray(value)

const isJsonMediaType = (contentType: string | null) =>
  contentType !== null &&
  /^application\/(?:[a-z0-9!#$&^_.+-]+\+)?json(?:\s*;|$)/i.test(contentType)

export const isReward = (value: unknown): value is Reward => {
  if (!isRecord(value) || Object.keys(value).length !== 6) {
    return false
  }
  for (const key of ['id', 'name', 'points', 'is_active', 'created_at', 'updated_at']) {
    if (!Object.hasOwn(value, key)) {
      return false
    }
  }

  return (
    typeof value.id === 'number' &&
    Number.isInteger(value.id) &&
    typeof value.name === 'string' &&
    typeof value.points === 'number' &&
    Number.isInteger(value.points) &&
    typeof value.is_active === 'boolean' &&
    typeof value.created_at === 'string' &&
    typeof value.updated_at === 'string'
  )
}

const isRewardList = (value: unknown): value is Reward[] =>
  Array.isArray(value) && value.every(isReward)

const readFieldErrors = (body: unknown): Record<string, string> => {
  if (!isRecord(body)) {
    return {}
  }

  const fieldErrors: Record<string, string> = {}
  for (const [field, messages] of Object.entries(body)) {
    if (
      Array.isArray(messages) &&
      messages.length > 0 &&
      typeof messages[0] === 'string'
    ) {
      fieldErrors[field] = messages[0]
    }
  }
  return fieldErrors
}

const readJsonBody = async (response: Response): Promise<unknown> => {
  if (!isJsonMediaType(response.headers.get('Content-Type'))) {
    return undefined
  }
  try {
    return await response.json()
  } catch {
    return undefined
  }
}

const send = async (path: string, init: RequestInit): Promise<Response> => {
  try {
    return await fetch(path, { credentials: 'same-origin', ...init })
  } catch {
    throw new RewardRequestError('unavailable')
  }
}

const readRewardBody = async (response: Response): Promise<Reward> => {
  if (response.redirected) {
    throw new RewardRequestError('unavailable')
  }
  const body = await readJsonBody(response)

  if (response.status === 400) {
    throw new RewardRequestError('validation', readFieldErrors(body))
  }
  if (response.status === 403) {
    throw new RewardRequestError('forbidden')
  }
  if (response.status === 404) {
    throw new RewardRequestError('notFound')
  }
  if ((response.status !== 200 && response.status !== 201) || !isReward(body)) {
    throw new RewardRequestError('unavailable')
  }
  return body
}

export interface ChildReward {
  id: number
  name: string
  points: number
}

export const isChildReward = (value: unknown): value is ChildReward => {
  if (!isRecord(value) || Object.keys(value).length !== 3) {
    return false
  }
  for (const key of ['id', 'name', 'points']) {
    if (!Object.hasOwn(value, key)) {
      return false
    }
  }
  return (
    typeof value.id === 'number' &&
    Number.isInteger(value.id) &&
    typeof value.name === 'string' &&
    typeof value.points === 'number' &&
    Number.isInteger(value.points)
  )
}

const isChildRewardList = (value: unknown): value is ChildReward[] =>
  Array.isArray(value) && value.every(isChildReward)

export const CHILD_REWARDS_QUERY_KEY = ['rewards', 'child'] as const

/** `GET /api/v1/rewards/`: the active catalogue, in the child's own shape. */
export const fetchChildRewards = async ({
  signal,
}: {
  signal?: AbortSignal
}): Promise<ChildReward[]> => {
  const response = await send('/api/v1/rewards/', {
    method: 'GET',
    headers: { Accept: 'application/json' },
    signal,
  })

  if (response.redirected) {
    throw new RewardRequestError('unavailable')
  }
  if (response.status === 403) {
    throw new RewardRequestError('forbidden')
  }
  const body = await readJsonBody(response)
  if (response.status !== 200 || !isChildRewardList(body)) {
    throw new RewardRequestError('unavailable')
  }
  return body
}

/** `GET /api/v1/rewards/`: the catalogue, active-only unless `includeInactive`. */
export const fetchRewards = async ({
  includeInactive,
  signal,
}: {
  includeInactive: boolean
  signal?: AbortSignal
}): Promise<Reward[]> => {
  const path = includeInactive
    ? '/api/v1/rewards/?include_inactive=true'
    : '/api/v1/rewards/'
  const response = await send(path, {
    method: 'GET',
    headers: { Accept: 'application/json' },
    signal,
  })

  if (response.redirected) {
    throw new RewardRequestError('unavailable')
  }
  if (response.status === 403) {
    throw new RewardRequestError('forbidden')
  }
  const body = await readJsonBody(response)
  if (response.status !== 200 || !isRewardList(body)) {
    throw new RewardRequestError('unavailable')
  }
  return body
}

interface WriteArgs {
  csrfToken: string
  signal?: AbortSignal
}

/** `POST /api/v1/rewards/`: create one reward. */
export const createReward = async ({
  csrfToken,
  name,
  points,
  signal,
}: WriteArgs & { name: string; points: number }): Promise<Reward> =>
  readRewardBody(
    await send('/api/v1/rewards/', {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        [CSRF_HEADER_NAME]: csrfToken,
      },
      body: JSON.stringify({ name, points }),
      signal,
    }),
  )

/** `PATCH /api/v1/rewards/<id>/`: edit an existing reward's name and points. */
export const updateReward = async ({
  csrfToken,
  id,
  name,
  points,
  signal,
}: WriteArgs & { id: number; name: string; points: number }): Promise<Reward> =>
  readRewardBody(
    await send(`/api/v1/rewards/${id}/`, {
      method: 'PATCH',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        [CSRF_HEADER_NAME]: csrfToken,
      },
      body: JSON.stringify({ name, points }),
      signal,
    }),
  )

/** `DELETE /api/v1/rewards/<id>/`: remove a reward permanently. 204 is success. */
export const deleteReward = async ({
  csrfToken,
  id,
  signal,
}: WriteArgs & { id: number }): Promise<void> => {
  const response = await send(`/api/v1/rewards/${id}/`, {
    method: 'DELETE',
    headers: {
      Accept: 'application/json',
      [CSRF_HEADER_NAME]: csrfToken,
    },
    signal,
  })

  if (response.status === 204) {
    return
  }
  if (response.status === 403) {
    throw new RewardRequestError('forbidden')
  }
  if (response.status === 404) {
    throw new RewardRequestError('notFound')
  }
  throw new RewardRequestError('unavailable')
}

const postStateChange = async (
  path: string,
  { csrfToken, signal }: WriteArgs,
): Promise<Reward> =>
  readRewardBody(
    await send(path, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        [CSRF_HEADER_NAME]: csrfToken,
      },
      signal,
    }),
  )

/** `POST /api/v1/rewards/<id>/deactivate/`: take a reward out of use. */
export const deactivateReward = async (
  args: WriteArgs & { id: number },
): Promise<Reward> => postStateChange(`/api/v1/rewards/${args.id}/deactivate/`, args)

/** `POST /api/v1/rewards/<id>/reactivate/`: put a reward back into use. */
export const reactivateReward = async (
  args: WriteArgs & { id: number },
): Promise<Reward> => postStateChange(`/api/v1/rewards/${args.id}/reactivate/`, args)

export { ensureCsrfToken }
